File: frontend/src/types/runner.ts
import { ItemWithResourceID } from "@/hooks/useResourceForItems";
import { Image } from "./images";
import { Machine } from "./machines";

// Define runner states
export type RunnerState = "starting" | "ready" | "awaiting_client" | "active" | "terminated" | "runner_starting" | "ready_claimed" | "closed" | "closed_pool";

// Define the frontend runner interface (camelCase for frontend use)
export interface Runner extends ItemWithResourceID<number> {
  id: number;
  userId?: number; // Optional because runners in the pool don't have a user yet
  image?: Image;
  imageId: number;
  machine?: Machine;
  machineId: number;
  keyId: number;
  state: RunnerState;
  identifier: string;
  externalHash: string;
  url?: string; // Will be populated when the runner is ready
  userIP?: string;
  envData?: JSON;
  sessionStart?: string; // When user requests runner
  sessionEnd?: string; // When runner expires
  endedOn?: string;
  createdOn?: string;
  updatedOn?: string;
  modifiedBy?: string;
  terminalToken?: string; // Token for the terminal session
  userEmail?: string; // Optional, can be used for user identification
}

export interface RunnerResponse{
  id: number,
  machine_id: number;
  image_id: number,
  user_id: number,
  key_id: number,
  state: RunnerState,
  url: string,
  user_ip: string,
  identifier: string,
  external_hash: string,
  env_data: JSON,
  session_start: string,
  session_end: string,
  ended_on: string,
  created_on: string,
  updated_on: string,
  modified_by: string,
  created_by: string,
  terminal_token: string,
  lifecycle_token: string,
  user_email?: string,
}


export const convertRunnerResponse = (runnerResponse: RunnerResponse): Runner => {
  if (!runnerResponse || typeof runnerResponse !== 'object') {
    throw new Error('Invalid machine data provided to converter');
  }

   return {
        id: runnerResponse.id,
        userId: runnerResponse.user_id,
        imageId: runnerResponse.image_id,
        machineId: runnerResponse.machine_id,
        keyId: runnerResponse.key_id,
        state: runnerResponse.state,
        identifier: runnerResponse.identifier,
        externalHash: runnerResponse.external_hash,
        url: runnerResponse.url,
        userIP: runnerResponse.user_ip,
        envData: runnerResponse.env_data,
        sessionStart: new Date(runnerResponse.session_start).toLocaleString('en-US'),
        sessionEnd: new Date(runnerResponse.session_end).toLocaleString('en-US'),
        endedOn: new Date(runnerResponse.ended_on).toLocaleString('en-US'),
        createdOn: new Date(runnerResponse.created_on).toLocaleDateString('en-US'),
        updatedOn: new Date(runnerResponse.updated_on).toLocaleDateString('en-US'),
        modifiedBy: runnerResponse.modified_by,
        terminalToken: runnerResponse.terminal_token,
        userEmail: runnerResponse.user_email,
      };
}

File: backend/app/models/runner.py
"""Runner model."""

from __future__ import annotations
from typing import Any
from datetime import datetime
from sqlmodel import SQLModel, Field, Session
from sqlalchemy import Column, JSON
from app.models.mixins import TimestampMixin
from pydantic import BaseModel
from app.db.database import engine

# states
# runner_starting
# app_starting
# ready
# setup
# awaiting_client
# active
# disconnecting
# disconnected
# closed
# terminated

# runner_alive states = [runner_starting, app_starting, ready, setup, awaiting_client, active, disconnecting, disconnected]
# runner_dead states = [closed, terminated]

# Relationships
# machine: Mapped["Machine"] = Relationship(back_populates="runners")
# image: Mapped["Image"] = Relationship(back_populates="runners")
# user: Mapped["User"] = Relationship(back_populates="runners")
# runner_histories: Mapped[List["RunnerHistory"]] = Relationship(back_populates="runner")

class Runner(TimestampMixin, SQLModel, table=True):
    """Runner model for the application."""

    id: int | None = Field(default=None, primary_key=True)
    machine_id: int = Field(foreign_key="machine.id")
    image_id: int = Field(foreign_key="image.id")
    user_id: int | None = Field(default=None, foreign_key="user.id")
    key_id: int | None = Field(default=None, foreign_key="key.id")
    state: str
    url: str
    lifecycle_token: str | None = None
    terminal_token: str | None = None
    user_ip: str | None = None
    identifier: str
    external_hash: str
    env_data: dict[str, Any] = Field(
        default={},
        sa_column=Column(JSON, nullable=False)
    )
    session_start: datetime | None = None
    session_end: datetime | None = None
    ended_on: datetime | None = None

    @property
    def is_alive_state(self) -> bool:
        """Return True if the runner's state is considered 'alive'."""
        alive_states = {
            "runner_starting", "app_starting", "ready", "runner_starting_claimed",
            "ready_claimed", "setup",
            "awaiting_client", "active", "disconnecting", "disconnected"
        }
        return self.state in alive_states

    @property
    def should_run_terminate_script(self) -> bool:
        """Return True if the runner's state is considered 'in use'."""
        do_not_run = {
            "ready", "ready_claimed", "runner_starting_claimed",
            "runner_starting", "app_starting", "terminated", "closed"
        }
        return self.state not in do_not_run

class RunnerResponse(BaseModel):
    """Runner response model with user email."""

    id: int
    machine_id: int
    image_id: int
    user_id: int | None = None
    key_id: int | None = None
    state: str
    url: str
    lifecycle_token: str | None = None
    terminal_token: str | None = None
    user_ip: str | None = None
    identifier: str
    external_hash: str
    env_data: dict[str, Any]
    session_start: datetime | None = None
    session_end: datetime | None = None
    ended_on: datetime | None = None

    # Add timestamp fields from TimestampMixin
    created_on: datetime | None = None
    updated_on: datetime | None = None
    modified_by: str | None = None
    created_by: str | None = None

    # Add user email field
    user_email: str | None = None

    class Config:
        """config for runner response model."""

        orm_mode = True

class RunnerUpdate(TimestampMixin, SQLModel):
    """Runner update model."""

    id: int
    state: str
    url: str
    user_ip: str | None = None
    external_hash: str
    env_data: dict[str, Any] | None = None
    session_start: datetime | None = None
    session_end: datetime | None = None
    ended_on: datetime | None = None

def create_runner(runner: Runner):
    """Create a runner record in the database."""
    with Session(engine) as session:
        session.add(runner)
        session.commit()
        session.refresh()
    return runner

def update_runner(runner: RunnerUpdate):
    """Update a runner record in the database."""
    with Session(engine) as session:
        runner_from_db = session.get(Runner, runner.id)
        runner_data = runner.model_dump(exclude_unset=True)
        runner_from_db.sqlmodel_update(runner_data)
        session.add(runner_from_db)
        session.commit()
        session.refresh(runner_from_db)
        return runner_from_db


def get_runner(runner_id: int):
    """Get a runner record from the database."""
    with Session(engine) as session:
        return session.get(Runner, runner_id)

def delete_runner(runner_id: int):
    """Delete a runner record from the database."""
    with Session(engine) as session:
        session.delete(runner_id)
        session.commit()
